### ISSUE TITLE
TensorRT batch inference

### ISSUE THREAD

**Issue by erdongchendou** [ultralytics/ultralytics#7100]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I export the yolov8 detection model with the following code.
``` python 
import os
from ultralytics import YOLO

model = YOLO("./yolov8l.pt")
model.export(format="engine", imgsz=640, device="0", dynamic=True, batch=6)
```

Do inference with the following code.

``` python 
import torch

model = YOLO("./yolov8l.engine") 
source = torch.rand(6, 3, 640, 640, dtype=torch.float32)
results = model.predict(source, save=False, batch=6, device="0")
```

And I get the following log, it seems like trt yolov8 does not do batch inference but do inference on one image a time.

``` bash
0: 640x640 (no detections), 1: 640x640 (no detections), 2: 640x640 (no detections), 
3: 640x640 (no detections), 4: 640x640 (no detections), 5: 640x640 (no detections), 
194.0ms
Speed: 0.5ms preprocess, 32.3ms inference, 6.6ms postprocess per image at shape (1, 3, 640, 640)
```

Do I do something wrong?

**Comment by glenn-jocher** [comment]
@erdongchendou hello! Thank you for providing detailed information about the issue you're facing with TensorRT batch inference. 🤔

From what you've described, it does indeed look like the inference is proceeding image by image instead of using batch processing. This could stem from a limitation or configuration issue within the TensorRT settings related to dynamic batching.

To resolve this, please ensure that the model export with TensorRT was successful with the specified batch size. Additionally, verify that the TensorRT execution context is appropriately set up for batch inference when you load the `yolov8l.engine` model.

If the settings seem correct but the problem persists, it might be a bug or an unimplemented feature. In that case, your detailed description would be invaluable in addressing the issue. Please raise an issue on the Ultralytics YOLOv8 repo detailing your findings, and we will look into it promptly.

For guided assistance on model exporting and predict functions, feel free to consult our documentation, which includes more information on advanced configuration and troubleshooting. If there's anything more I can assist you with, let me know! 👍

**Comment by erdongchendou** [comment]
@glenn-jocher I think the export process is successful, the following is the export log.
> Ultralytics YOLOv8.0.219 🚀 Python-3.9.18 torch-1.8.2+cu111 CUDA:0 (Tesla T4, 15110MiB)
YOLOv8l summary (fused): 268 layers, 43668288 parameters, 0 gradients, 165.2 GFLOPs
> 
> PyTorch: starting from '/hmty/ultralytics/yolov8l.pt' with input shape (6, 3, 640, 640) BCHW and output shape(s) (6, 84, 8400) (83.7 MB)
> 
> ONNX: starting export with onnx 1.15.0 opset 12...
> /usr/local/lib/python3.9/site-packages/torch/tensor.py:587: RuntimeWarning: Iterating over a tensor might cause the trace to be incorrect. Passing a tensor of different shape won't change the number of iterations executed (and might lead to errors or silently give incorrect results).
>   warnings.warn('Iterating over a tensor might cause the trace to be incorrect. '
> ONNX: simplifying with onnxsim 0.4.35...
> ONNX: export success ✅ 9.9s, saved as '/hmty/ultralytics/yolov8l.onnx' (166.6 MB)
> 
> TensorRT: starting export with TensorRT 8.4.3.1...
> [12/22/2023-01:24:56] [TRT] [I] [MemUsageChange] Init CUDA: CPU +302, GPU +0, now: CPU 3645, GPU 6698 (MiB)
> [12/22/2023-01:24:57] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +206, GPU +72, now: CPU 3871, GPU 6770 (MiB)
> [12/22/2023-01:24:57] [TRT] [I] ----------------------------------------------------------------
> [12/22/2023-01:24:57] [TRT] [I] Input filename:   /hmty/ultralytics/yolov8l.onnx
> [12/22/2023-01:24:57] [TRT] [I] ONNX IR version:  0.0.6
> [12/22/2023-01:24:57] [TRT] [I] Opset version:    12
> [12/22/2023-01:24:57] [TRT] [I] Producer name:    pytorch
> [12/22/2023-01:24:57] [TRT] [I] Producer version: 1.8
> [12/22/2023-01:24:57] [TRT] [I] Domain:           
> [12/22/2023-01:24:57] [TRT] [I] Model version:    0
> [12/22/2023-01:24:57] [TRT] [I] Doc string:       
> [12/22/2023-01:24:57] [TRT] [I] ----------------------------------------------------------------
> [12/22/2023-01:24:57] [TRT] [W] onnx2trt_utils.cpp:369: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
> TensorRT: input "images" with shape(-1, 3, -1, -1) DataType.FLOAT
> TensorRT: output "output0" with shape(-1, 84, -1) DataType.FLOAT
> TensorRT: building FP32 engine as /hmty/ultralytics/yolov8l.engine
> [12/22/2023-01:24:59] [TRT] [I] [MemUsageChange] Init cuBLAS/cuBLASLt: CPU +512, GPU +244, now: CPU 4568, GPU 6362 (MiB)
> [12/22/2023-01:25:00] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +118, GPU +76, now: CPU 4686, GPU 6438 (MiB)
> [12/22/2023-01:25:00] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
> [12/22/2023-01:25:00] [TRT] [I] Local timing cache in use. Profiling results in this builder pass will not be stored.
> [12/22/2023-01:25:43] [TRT] [I] Some tactics do not have sufficient workspace memory to run. Increasing workspace size will enable more tactics, please check verbose output for requested sizes.
> 
> 
> [12/22/2023-01:27:19] [TRT] [I] Detected 1 inputs and 3 output network tensors.
> [12/22/2023-01:27:20] [TRT] [I] Total Host Persistent Memory: 147328
> [12/22/2023-01:27:20] [TRT] [I] Total Device Persistent Memory: 1532416
> [12/22/2023-01:27:20] [TRT] [I] Total Scratch Memory: 27070976
> [12/22/2023-01:27:20] [TRT] [I] [MemUsageStats] Peak memory usage of TRT CPU/GPU memory allocators: CPU 0 MiB, GPU 0 MiB
> [12/22/2023-01:27:20] [TRT] [I] [BlockAssignment] Algorithm ShiftNTopDown took 137.348ms to assign 13 blocks to 273 nodes requiring 579994120 bytes.
> [12/22/2023-01:27:20] [TRT] [I] Total Activation Memory: 579994120
> [12/22/2023-01:27:20] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +0, GPU +34, now: CPU 5205, GPU 6878 (MiB)
> [12/22/2023-01:27:20] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
> [12/22/2023-01:27:20] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in building engine: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
> [12/22/2023-01:27:20] [TRT] [W] The getMaxBatchSize() function should not be used with an engine built from a network created with NetworkDefinitionCreationFlag::kEXPLICIT_BATCH flag. This function will always return 1.
> [12/22/2023-01:27:20] [TRT] [W] The getMaxBatchSize() function should not be used with an engine built from a network created with NetworkDefinitionCreationFlag::kEXPLICIT_BATCH flag. This function will always return 1.
> TensorRT: export success ✅ 155.4s, saved as '/hmty/ultralytics/yolov8l.engine' (190.7 MB)
> 
> Export complete (161.7s)
> Results saved to /hmty/ultralytics
> Predict:         yolo predict task=detect model=/hmty/ultralytics/yolov8l.engine imgsz=640  
> Validate:        yolo val task=detect model=/hmty/ultralytics/yolov8l.engine imgsz=640 data=coco.yaml  
> Visualize:       https://netron.app


The following is the inference log.


> WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify', or 'pose'.
> Loading yolov8l.engine for TensorRT inference...
> [12/22/2023-01:31:52] [TRT] [I] [MemUsageChange] Init CUDA: CPU +310, GPU +0, now: CPU 407, GPU 4548 (MiB)
> [12/22/2023-01:31:52] [TRT] [I] Loaded engine size: 190 MiB
> [12/22/2023-01:31:53] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +629, GPU +292, now: CPU 1262, GPU 5034 (MiB)
> [12/22/2023-01:31:53] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
> [12/22/2023-01:31:53] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in engine deserialization: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
> [12/22/2023-01:31:53] [TRT] [I] [MemUsageChange] Init cuDNN: CPU +0, GPU +32, now: CPU 1071, GPU 5034 (MiB)
> [12/22/2023-01:31:53] [TRT] [W] TensorRT was linked against cuDNN 8.4.1 but loaded cuDNN 8.4.0
> [12/22/2023-01:31:53] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +0, now: CPU 0, GPU 0 (MiB)
> 
> 0: 640x640 (no detections), 1: 640x640 (no detections), 2: 640x640 (no detections), 3: 640x640 (no detections), 4: 640x640 (no detections), 5: 640x640 (no detections), 238.2ms
> Speed: 0.6ms preprocess, 39.7ms inference, 5.2ms postprocess per image at shape (1, 3, 640, 640)


The code I use is the example code from here: https://docs.ultralytics.com/modes/export/#arguments 
You said I need to ensure that the model export with TensorRT was successful with the specified batch size. Additionally, verify that the TensorRT execution context is appropriately set up for batch inference when you load the `yolov8l.engine` model.

From my experience and knowledge, I think the export process is successful and I set batch size at inference time. So I can not find where the problem is. 
From your experience, could you please point out where I did wrong？ And by the way, have you ever do batch inference with yolov8 TensorRT engine?

**Comment by glenn-jocher** [comment]
Hello @erdongchendou,

From the logs you have provided, it does appear that the export and loading processes for your TensorRT engine are functioning well with the expected shapes and batch sizes. The warning regarding the `getMaxBatchSize()` function not being applicable due to the use of explicit batch network definitions indicates the model should support dynamic batching.

However, since you are still observing image-by-image inference rather than full batch processing, it seems there could be an issue at inference time rather than export. As you have followed the documentation and prescribed practices, I suspect the issue could lie deeper within the TensorRT execution context or the handling of batched input within the Python `predict()` function.

As for your question about batch inference with the YOLOv8 TensorRT engine, at Ultralytics we do our due prescriptive checks to support such features, but the actual application may vary based on user environments and the specifics of the TensorRT version employed.

For further insight into this issue, we encourage you to open a detailed issue on the Ultralytics YOLOv8 repo, providing all the logs and steps to replicate your scenario. We prioritize such instances to ensure robustness across varied use cases and appreciate your effort in highlighting this matter.

Thank you for bringing this to our attention. Our team is committed to assisting you in resolving this issue efficiently.

**Comment by erdongchendou** [comment]
Start another issue as suggested.

**Comment by glenn-jocher** [comment]
@erdongchendou, thank you for the detailed information. It appears you've taken the correct steps in following our export and inference guidelines. If batch inference isn't working as expected despite this, I recommend opening a new issue on the Ultralytics YOLOv8 repo with all relevant logs and a step-by-step breakdown of your process. That will enable our team to replicate the scenario and delve deeper into the matter.

We appreciate your patience and are here to help get this resolved! 🛠️

**Comment by amani784** [comment]
Hello! I used batch processing to optimize person detection by analyzing video frames in groups of 32 instead of processing them one by one, passing them simultaneously to the YOLO model optimized with TensorRT. Each batch is processed in a single prediction call. However, using different batch sizes (8, 16, 32) does not improve the FPS or reduce the detection time per frame. Is this normal? If not, what could be the solution?

**Comment by glenn-jocher** [comment]
@amani784 This issue with batch processing not improving performance is consistent with what @erdongchendou reported. While TensorRT should theoretically improve throughput with batching, our current implementation may not be fully optimizing the batch inference path.

Some possible causes and solutions:
1. The Python-level predict() function may be serializing batch inference under the hood
2. Your GPU might be reaching its compute capacity regardless of batch size
3. Memory transfers between CPU and GPU could be creating bottlenecks

You might try:
- Verifying GPU utilization during inference (using nvidia-smi)
- Testing with lower precision (FP16 or INT8) by adding half=True or int8=True during export
- Adjusting workspace size parameter to provide more memory for optimization

For more detailed diagnostics, please include your complete export code, inference code, and hardware specifications in a new issue. This would help us replicate and address the specific behavior you're seeing.
